### PR TITLE
Improve FindDockerImageUses recipe with multistage files and gitlab-ci files

### PR DESCRIPTION
**Pull request by jevanlingen**
## What's changed?
Use the new `IMAGE` reference type to improve the `FindDockerImageUses` recipe. The updated recipes has proper support for multistage  dockerfiles and yaml files with an image key (for example a `.gitlab-ci` file).

## What's your motivation?
Run one find (and replace?) recipe to search for your docker images in one go, no matter how the image has been configured.

## Note
Notice the `IMAGE` reference type is also newly introduced in:

- https://github.com/openrewrite/rewrite/pull/4793

## Any additional context 
This work is a step-up for integration of more ImageProviders. For example, testcontainers is also a project which uses images.

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

**Comment by jevanlingen** [comment]
Notice build will fail until
- https://github.com/openrewrite/rewrite/pull/4793

will be merged!

File: src/main/java/org/openrewrite/docker/table/DockerBaseImages.java [hunk 1]
```diff
@@ -30,6 +30,14 @@ public DockerBaseImages(Recipe recipe) {
 
     @Value
     public static class Row {
+        @Column(displayName = "Source path before the run",
+                description = "The source path of the file before the run.")
+        String sourcePath;
+
+        @Column(displayName = "LST type",
+                description = "The LST model type that the file is parsed as.")
+        String type;
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/docker/table/DockerBaseImages.java]: The name and description hinted at possibly changing the file path, which isn't something we'll support.
```suggestion
        @Column(displayName = "Source path",
                description = "The source file containing the image reference.")
        String sourcePath;
```

The LST type I'd consider an implementation detail, that might still change in the future when we add new parsers. I'd lean towards leaving that out.

File: src/main/java/org/openrewrite/docker/search/FindDockerImageUses.java [hunk 2]
```diff
@@ -37,25 +42,52 @@ public String getDisplayName() {
...
+            public @Nullable Tree visit(@Nullable Tree tree, ExecutionContext ctx) {
+                if (tree instanceof SourceFileWithReferences) {
+                    SourceFileWithReferences sourceFile = (SourceFileWithReferences) tree;
+                    Path sourcePath = sourceFile.getSourcePath();
+                    Collection<Reference> references = sourceFile.getReferences().findMatches(new ImageMatcher());
+                    Map<Tree, List<Reference>> matches = new HashMap<>();
+                    for (Reference ref : references) {
+                        DockerImageVersion from = DockerImageVersion.of(ref.getValue());
+                        dockerBaseImages.insertRow(ctx,
+                                new DockerBaseImages.Row(sourcePath.toString(), tree.getClass().getSimpleName(), from.getImageName(), from.getVersion())
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/docker/search/FindDockerImageUses.java]:
```suggestion
                                new DockerBaseImages.Row(sourcePath.toString(), from.getImageName(), from.getVersion())
```

File: src/main/java/org/openrewrite/docker/DockerImageVersion.java [hunk 3]
```diff
@@ -22,11 +22,10 @@
 public class DockerImageVersion {
     String imageName;
 
-    @Nullable
     String version;
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/docker/DockerImageVersion.java]: Why this particular change? I don't think an empty string is quite the same as indicating clearly there was no version.
You can see this reflected further down where we now need to determine if this was an empty string before printing `:`.

**Review by timtebeek — APPROVED** [review]
Approved already. Ready to merge once the final question is answered.

File: src/main/java/org/openrewrite/docker/DockerImageVersion.java (hunk 3 above)
Comment (jevanlingen) [reply src/main/java/org/openrewrite/docker/DockerImageVersion.java]: No real reason; me focussing on FP concepts the last couple of years let this slip in. I'll revert it!

_Sidenote: having the JSpecify stuff everywhere in our repo's, makes defining non-null object less of a need, even from an FP standpoint!!_

**Pull request merged**